### PR TITLE
change: New error handling APIs (merging v7 branch into master)

### LINKED ISSUE #354: Error handling refactor
We are in the process of implementing a large-scale refactor of our exception types. This is largely a breaking change, and therefore will be released in the next major version (no concrete timeline yet). The benefits of this refactor include:

1. Clearly defined top-level error codes for all exceptions, where error codes follow the specification in https://cloud.google.com/apis/design/errors.
2. Ability to provide both a top-level and service-specific error codes for some exceptions.
3. Ability to access HTTP response (including content and headers) via exceptions.
4. Clear separation between developer errors, and backend errors.

The overall API changes planned as part of this refactor are outlined below (taking FCM as an example):

```
class OutgoingHttpRequest {
  String getMethod()
  String getUrl()
  HttpContent getContent()
  Map<String, Object> getHeaders()
}

class IncomingHttpResponse {
  int getStatusCode()
  String getContent()
  Map<String, Object> getHeaders()
  OutgoingHttpRequest getRequest()
}

enum ErrorCode {
  INVALID_ARGUMENT,
  INTERNAL,
  UNAVAILABLE,
  // ... other top-level error codes
}

class FirebaseException extends Exception {
  ErrorCode getErrorCode();
  String getMessage();
  Throwable getCause();
  IncomingHttpResponse getHttpResponse()
}

enum MessagingErrorCode {
  UNREGISTERED,
  SENDER_ID_MISMATCH,
  // ... other FCM error codes
}

class FirebaseMessagingException extends FirebaseException {
  MessagingErrorCode getMessagingErrorCode();
}
```

### PR DESCRIPTION
**Pull request by hiranya911**
RELEASE NOTE: feat: Added a collection of new APIs for implementing advanced error handling logic. See [error handling guide](https://firebase.google.com/docs/reference/admin/error-handling) for more details.
RELEASE NOTE: feat: Added a new `ErrorCode` enum that defines all platform error codes.
RELEASE NOTE: feat: Added a new `IncomingHttpResponse` class that can be used to access the HTTP response object associated with an exception.

RELEASE NOTE: feat(auth): Added a new `AuthErrorCode` enum that defines all Firebase Auth error codes.
RELEASE NOTE: change(auth): `FirebaseAuthException` now exposes its error code as an `AuthErrorCode` value.
RELEASE NOTE: change(auth): All implementations of the `UserImportHash` abstract class are now marked `final`.

RELEASE NOTE: feat(fcm): Added a new `MessagingErrorCode` enum that defines all Firebase Cloud Messaging error codes.
RELEASE NOTE: change(fcm): `FirebaseMessagingException` now exposes its error code as a `MessagingErrorCode` value.
RELEASE NOTE: change(fcm): Public constructors of the `Notification` class have been removed. Use the `Notification.builder()` to create new instances.

Resolves #354

### CODE REVIEW

**Comment by hiranya911** [comment]
@egilmorez can you take a look at the API doc portions in the following classes?

* ErrorCode
* AuthErrorCode
* MessagingErrorCode
* FirebaseException
* FirebaseAuthException
* FirebaseMessagingException
* IncomingHttpResponse
* OutgoingHttpRequest

**Comment by hiranya911** [comment]
@kevinthecheung adding you to this as well since you're familiar with the Auth API surface. Appreciate if you can review the API docs for the classes listed in my previous comment.

File: src/main/java/com/google/firebase/ErrorCode.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  /**
+   * Client does not have sufficient permission. This can happen because the OAuth token does
+   * not have the right scopes, the client doesn't have permission, or the API has not been
+   * enabled for the client project.
+   */
+  PERMISSION_DENIED,
+
+  /**
+   * A specified resource is not found, or the request is rejected by undisclosed reasons,
+   * such as whitelisting.
```
Comment (kevinthecheung) [inline src/main/java/com/google/firebase/ErrorCode.java:54]:
```suggestion
   * A specified resource is not found, or the request is rejected for unknown reasons,
   * such as a blocked network address.
```

File: src/main/java/com/google/firebase/ErrorCode.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+   */
+  NOT_FOUND,
+
+  /**
+   * Concurrency conflict, such as read-modify-write conflict.
+   */
+  CONFLICT,
+
+  /**
+   * Concurrency conflict, such as read-modify-write conflict.
```
Comment (kevinthecheung) [inline src/main/java/com/google/firebase/ErrorCode.java:64]: Intentional dupe?

File: src/main/java/com/google/firebase/ErrorCode.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+   */
+  ABORTED,
+
+  /**
+   * The resource that a client tried to create already exists.
+   */
+  ALREADY_EXISTS,
+
+  /**
+   * Either out of resource quota or reaching rate limiting.
```
Comment (kevinthecheung) [inline src/main/java/com/google/firebase/ErrorCode.java:74]:
```suggestion
   * Either out of resource quota or rate limited.
```

File: src/main/java/com/google/firebase/ErrorCode.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+public enum ErrorCode {
+
+  /**
+   * Client specified an invalid argument.
+   */
+  INVALID_ARGUMENT,
+
+  /**
+   * Request can not be executed in the current system state, such as deleting a non-empty
+   * directory.
```
Comment (egilmorez) [inline src/main/java/com/google/firebase/ErrorCode.java:31]:
```suggestion
  /**
   * Request cannot be executed in the current system state, such as deleting a non-empty
   * directory.
```

File: src/main/java/com/google/firebase/ErrorCode.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  INTERNAL,
+
+  /**
+   * Service unavailable. Typically the server is down.
+   */
+  UNAVAILABLE,
+
+  /**
+   * Request deadline exceeded. This will happen only if the caller sets a deadline that is
+   * shorter than the method's default deadline (i.e. requested deadline is not enough for the
```
Comment (egilmorez) [inline src/main/java/com/google/firebase/ErrorCode.java:105]:
```suggestion
  /**
   * Request deadline exceeded. This happens only if the caller sets a deadline that is
   * shorter than the method's default deadline (i.e. requested deadline is not enough for the
```

File: src/main/java/com/google/firebase/FirebaseOptions.java [hunk 6]
```diff
@@ -223,6 +223,16 @@ public static Builder builder() {
     return new Builder();
   }
 
+  /**
+   * Creates a new Builder from the options object.
```
Comment (egilmorez) [inline src/main/java/com/google/firebase/FirebaseOptions.java:227]: Should "Builder" be a literal?

File: src/main/java/com/google/firebase/FirebaseOptions.java [hunk 7]
```diff
@@ -223,6 +223,16 @@ public static Builder builder() {
...
+   *
+   * <p>The new builder is not backed by this object's values, that is changes made to the new
+   * builder don't change the values of the origin object.
+   */
```
Comment (egilmorez) [inline src/main/java/com/google/firebase/FirebaseOptions.java:231]:
```suggestion
   * <p>The new builder is not backed by this object's values; that is, changes made to the new
   * builder don't change the values of the origin object.
   */
```

File: src/main/java/com/google/firebase/IncomingHttpResponse.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.http.HttpResponseException;
+import com.google.common.collect.ImmutableMap;
+import com.google.firebase.database.annotations.Nullable;
+import java.util.Map;
+
+/**
+ * Contains the information that describe an HTTP response received by the SDK.
```
Comment (egilmorez) [inline src/main/java/com/google/firebase/IncomingHttpResponse.java:29]:
```suggestion
 * Contains information that describes an HTTP response received by the SDK.
```

File: src/main/java/com/google/firebase/IncomingHttpResponse.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+ */
+public final class IncomingHttpResponse {
+
+  private final int statusCode;
+  private final String content;
+  private final Map<String, Object> headers;
+  private final OutgoingHttpRequest request;
+
+  /**
+   * Creates an IncomingHttpResponse from a successful HttpResponse and the content read from it.
```
Comment (egilmorez) [inline src/main/java/com/google/firebase/IncomingHttpResponse.java:39]:
```suggestion
   * Creates an `IncomingHttpResponse` from a successful `HttpResponse` and the content read from it.
```

File: src/main/java/com/google/firebase/IncomingHttpResponse.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+   * The caller is expected to read the content from the response, and handle any errors that
+   * may occur while reading.
+   *
+   * @param response A successful HttpResponse.
```
Comment (egilmorez) [inline src/main/java/com/google/firebase/IncomingHttpResponse.java:43]:
```suggestion
   * @param response A successful `HttpResponse`.
```

File: src/main/java/com/google/firebase/IncomingHttpResponse.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+  public IncomingHttpResponse(HttpResponse response, @Nullable String content) {
+    checkNotNull(response, "response must not be null");
+    this.statusCode = response.getStatusCode();
+    this.content = content;
+    this.headers = ImmutableMap.copyOf(response.getHeaders());
+    this.request = new OutgoingHttpRequest(response.getRequest());
+  }
+
+  /**
+   * Creates an IncomingHttpResponse from an HTTP error response.
```
Comment (egilmorez) [inline src/main/java/com/google/firebase/IncomingHttpResponse.java:55]:
```suggestion
  /**
   * Creates an `IncomingHttpResponse` from an HTTP error response.
```

File: src/main/java/com/google/firebase/IncomingHttpResponse.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+   *
+   * @param e HttpResponseException representing the HTTP error response.
```
Comment (egilmorez) [inline src/main/java/com/google/firebase/IncomingHttpResponse.java:57]: In cases like this, do you think it's best to say "`HttpResponseException` object" ?

File: src/main/java/com/google/firebase/IncomingHttpResponse.java [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+   * @param request The HttpRequest that resulted in the error.
+   */
+  public IncomingHttpResponse(HttpResponseException e, HttpRequest request) {
+    this(e, new OutgoingHttpRequest(request));
+  }
+
+  /**
+   * Creates an IncomingHttpResponse from an HTTP error response.
```
Comment (egilmorez) [inline src/main/java/com/google/firebase/IncomingHttpResponse.java:65]:
```suggestion
  /**
   * Creates an `IncomingHttpResponse` from an HTTP error response.
```

File: src/main/java/com/google/firebase/OutgoingHttpRequest.java [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+ */
+public final class OutgoingHttpRequest {
+
+  private final String method;
+  private final String url;
+  private final HttpContent content;
+  private final Map<String, Object> headers;
+
+  /**
+   * Creates an OutgoingHttpRequest from the HTTP method and URL.
```
Comment (egilmorez) [inline src/main/java/com/google/firebase/OutgoingHttpRequest.java:40]:
```suggestion
  /**
   * Creates an `OutgoingHttpRequest` from the HTTP method and URL.
```

File: src/main/java/com/google/firebase/messaging/MessagingErrorCode.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+public enum MessagingErrorCode {
+
+  /**
+   * APNs certificate or web push auth key was invalid or missing.
+   */
+  THIRD_PARTY_AUTH_ERROR,
+
+  /**
+   * One or more argument specified in the request was invalid.
+   */
```
Comment (egilmorez) [inline src/main/java/com/google/firebase/messaging/MessagingErrorCode.java:15]:
```suggestion
  /**
   * One or more arguments specified in the request were invalid.
   */
```

File: src/main/java/com/google/firebase/ErrorCode.java (hunk 2 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/ErrorCode.java:64]: Yes. These 2 error codes are identical in semantics. We should be able to fold them into one in a future major release.

File: src/main/java/com/google/firebase/OutgoingHttpRequest.java (hunk 14 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/OutgoingHttpRequest.java]: Java requires `{@code }` syntax. Fixed manually.

File: src/main/java/com/google/firebase/IncomingHttpResponse.java (hunk 13 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/IncomingHttpResponse.java]: Fixed

File: src/main/java/com/google/firebase/IncomingHttpResponse.java (hunk 12 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/IncomingHttpResponse.java]: Changed to "Exception representing the..."

File: src/main/java/com/google/firebase/IncomingHttpResponse.java (hunk 11 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/IncomingHttpResponse.java]: Fixed

File: src/main/java/com/google/firebase/IncomingHttpResponse.java (hunk 10 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/IncomingHttpResponse.java]: Changed to "A successful response."

File: src/main/java/com/google/firebase/FirebaseOptions.java (hunk 6 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/FirebaseOptions.java]: Fixed

File: src/main/java/com/google/firebase/IncomingHttpResponse.java (hunk 9 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/IncomingHttpResponse.java]: Fixed

**Comment by hiranya911** [comment]
Thanks @egilmorez. I accepted most of the suggestions as is. For code-formatting type names we have to use the `{@code}` tag in Java, which I've applied manually.

**Pull request merged**